Unified PDF: Wasting time under dictionaryPopupInfoForSelection generating an unused attributed string

hortont424 opened the pull request:
#### bec768823d9a21850357aa34faec406ea6325d58
<pre>
Unified PDF: Wasting time under dictionaryPopupInfoForSelection generating an unused attributed string
<a href="https://bugs.webkit.org/show_bug.cgi?id=278151">https://bugs.webkit.org/show_bug.cgi?id=278151</a>
<a href="https://rdar.apple.com/133914757">rdar://133914757</a>

Reviewed by Abrar Rahman Protyasha.

Unified PDF uses proper TextIndicators (with snapshots and all) for Lookup.
Legacy PDFPlugin still uses attributed strings presented by Lookup itself.
Because of this, bifurcate our implementations of dictionaryPopupInfoForSelection.
Simplify Unified PDF&apos;s implementation; it can avoid doing any style computation, and
can avoid laying out the string (which was egregiously expensive for large selections).

Also leave a note for the future: once legacy PDFPlugin is gone,
DictionaryPopupInfo only needs a textual string, no more attributes!

* Source/WebCore/editing/DictionaryPopupInfo.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::dictionaryPopupInfoForSelection):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::performImmediateActionHitTestAtLocation):
(WebKit::PDFPluginBase::dictionaryPopupInfoForSelection): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::dictionaryPopupInfoForSelection):

Canonical link: <a href="https://commits.webkit.org/282276@main">https://commits.webkit.org/282276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/338e285e443740698bbdc8ddfe57e24ca6ea25a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42034 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15273 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66698 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13282 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64798 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49720 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13566 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50586 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9178 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65747 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39104 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54306 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31265 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35813 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11629 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12158 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57362 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11960 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68393 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6624 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57902 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6654 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54353 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58087 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13906 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5547 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37854 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/38934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40045 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1372,6 +1373,56 @@ static bool getEventTypeFromWebEvent(const WebEvent& event, NSEventType& eventTy
+        CocoaFont *font = [scaledAttributes objectForKey:NSFontAttributeName];
+        if (font) {
+            auto desiredFontSize = font.pointSize * scaleFactor;
+#if USE(APPKIT)
+            font = [[NSFontManager sharedFontManager] convertFont:font toSize:desiredFontSize];
+#else
+            font = [font fontWithSize:desiredFontSize];
+#endif
```
This file is `PLATFORM(MAC)` only, so we don't have to be so considerate! But also, no harm in keeping this around.

hortont424 committed:
Unified PDF: Wasting time under dictionaryPopupInfoForSelection generating an unused attributed string
https://bugs.webkit.org/show_bug.cgi?id=278151
rdar://133914757

Reviewed by Abrar Rahman Protyasha.

Unified PDF uses proper TextIndicators (with snapshots and all) for Lookup.
Legacy PDFPlugin still uses attributed strings presented by Lookup itself.
Because of this, bifurcate our implementations of dictionaryPopupInfoForSelection.
Simplify Unified PDF's implementation; it can avoid doing any style computation, and
can avoid laying out the string (which was egregiously expensive for large selections).

Also leave a note for the future: once legacy PDFPlugin is gone,
DictionaryPopupInfo only needs a textual string, no more attributes!

* Source/WebCore/editing/DictionaryPopupInfo.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::dictionaryPopupInfoForSelection):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::performImmediateActionHitTestAtLocation):
(WebKit::PDFPluginBase::dictionaryPopupInfoForSelection): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::dictionaryPopupInfoForSelection):

Canonical link: https://commits.webkit.org/282276@main
